src/query/v2/multiframe.cpp

```diff
@@ -0,0 +1,120 @@
+#include "utils/pmr/vector.hpp"
+
+namespace memgraph::query::v2 {
+
+MultiFrame::MultiFrame(FrameWithValidity default_frame, size_t number_of_frames,
+                       utils::MemoryResource *execution_memory)
+    : default_frame_(default_frame),
+      frames_(utils::pmr::vector<FrameWithValidity>(number_of_frames, default_frame, execution_memory)) {
+  MG_ASSERT(number_of_frames > 0);
+  MG_ASSERT(!default_frame.IsValid());
+}
+
+MultiFrame::MultiFrame(const MultiFrame &other) : default_frame_(other.default_frame_) {
+  /*
+  TODO
+  Do we just copy all frames or do we make distinctions between valid and not valid frames? Does it make any
+  difference?
+  */
+  frames_.reserve(other.frames_.size());
+  std::transform(other.frames_.begin(), other.frames_.end(), std::back_inserter(frames_),
+                 [&default_frame = default_frame_](const auto &other_frame) {
+                   if (other_frame.IsValid()) {
+                     return other_frame;
+                   }
+                   return default_frame;
+                 });
+}
+
+// NOLINTNEXTLINE (bugprone-exception-escape)
+MultiFrame::MultiFrame(MultiFrame &&other) noexcept
+    : default_frame_(std::move(other.default_frame_)), frames_(std::move(other.frames_)) {}
+
+void MultiFrame::ResetAllFramesInvalid() noexcept {
+  std::for_each(frames_.begin(), frames_.end(), [](auto &frame) { frame.MakeInvalid(); });
+}
+
+bool MultiFrame::HasValidFrame() const noexcept {
+  return std::any_of(frames_.begin(), frames_.end(), [](auto &frame) { return frame.IsValid(); });
+}
+
+// NOLINTNEXTLINE (bugprone-exception-escape)
+void MultiFrame::DefragmentValidFrames() noexcept {
+  /*
+  from: https://en.cppreference.com/w/cpp/algorithm/remove
+  "Removing is done by shifting (by means of copy assignment (until C++11)move assignment (since C++11)) the elements
+  in the range in such a way that the elements that are not to be removed appear in the beginning of the range.
+  Relative order of the elements that remain is preserved and the physical size of the container is unchanged."
+  */
+  [[maybe_unused]] const auto it =
+      std::remove_if(frames_.begin(), frames_.end(), [](auto &frame) { return !frame.IsValid(); });
+}
+
+ValidFramesReader MultiFrame::GetValidFramesReader() { return ValidFramesReader{*this}; }
+
+ValidFramesModifier MultiFrame::GetValidFramesModifier() { return ValidFramesModifier{*this}; }
+
+ValidFramesConsumer MultiFrame::GetValidFramesConsumer() { return ValidFramesConsumer{*this}; }
+
+InvalidFramesPopulator MultiFrame::GetInvalidFramesPopulator() { return InvalidFramesPopulator{*this}; }
+
+ValidFramesReader::ValidFramesReader(MultiFrame &multiframe) : multiframe_(multiframe) {}
+
+ValidFramesReader::Iterator ValidFramesReader::begin() { return Iterator{&multiframe_.frames_[0], *this}; }
+ValidFramesReader::Iterator ValidFramesReader::end() {
+  return Iterator{&multiframe_.frames_[multiframe_.frames_.size()], *this};
```

Comment:
Here you basically get the address of an object referenced by an invalid reference (because you index into the vector with `size()`). I would instead do `multiframe_.frames_.data() + multiframe_.frames_.size()`. This way it is totally fine until the pointer is not dereferenced. Of course, if you apply my suggestion about how to construct `ValidFramesReader` this comment might become obsolete for this iterator, but it remains true for others.